zipkin-storage/zipkin2_cassandra/src/main/java/zipkin2/storage/cassandra/InsertSpan.java

```diff
@@ -0,0 +1,184 @@
+
+    abstract Map<String, String> tags();
+
+    @Nullable abstract String annotation_query();
+
+    abstract boolean shared();
+
+    abstract boolean debug();
+  }
+
+  static class Factory {
+    final Session session;
+    final PreparedStatement preparedStatement;
+    final boolean strictTraceId;
+
+    Factory(Session session, boolean strictTraceId) {
+      this.session = session;
+      this.preparedStatement = session.prepare(QueryBuilder.insertInto(TABLE_SPAN)
+        .value("trace_id", QueryBuilder.bindMarker("trace_id"))
+        .value("trace_id_high", QueryBuilder.bindMarker("trace_id_high"))
+        .value("ts_uuid", QueryBuilder.bindMarker("ts_uuid"))
+        .value("parent_id", QueryBuilder.bindMarker("parent_id"))
+        .value("id", QueryBuilder.bindMarker("id"))
+        .value("kind", QueryBuilder.bindMarker("kind"))
+        .value("span", QueryBuilder.bindMarker("span"))
+        .value("ts", QueryBuilder.bindMarker("ts"))
+        .value("duration", QueryBuilder.bindMarker("duration"))
+        .value("l_ep", QueryBuilder.bindMarker("l_ep"))
+        .value("l_service", QueryBuilder.bindMarker("l_service"))
+        .value("r_ep", QueryBuilder.bindMarker("r_ep"))
+        .value("annotations", QueryBuilder.bindMarker("annotations"))
+        .value("tags", QueryBuilder.bindMarker("tags"))
+        .value("shared", QueryBuilder.bindMarker("shared"))
+        .value("debug", QueryBuilder.bindMarker("debug"))
+        .value("annotation_query", QueryBuilder.bindMarker("annotation_query")));
+      this.strictTraceId = strictTraceId;
+    }
+
+    Input newInput(zipkin2.Span span, UUID ts_uuid) {
+      boolean traceIdHigh = !strictTraceId && span.traceId().length() == 32;
+      List<AnnotationUDT> annotations;
+      if (!span.annotations().isEmpty()) {
+        annotations = span.annotations().stream()
+          .map(AnnotationUDT::new)
+          .collect(Collectors.toList());
+      } else {
+        annotations = Collections.emptyList();
+      }
+      String annotation_query = null;
+      Set<String> annotationQuery = CassandraUtil.annotationKeys(span);
+      if (!annotationQuery.isEmpty()) {
+        annotation_query = COMMA_JOINER.join(annotationQuery);
+      }
+      return new AutoValue_InsertSpan_Input(
+        ts_uuid,
+        traceIdHigh ? span.traceId().substring(0, 16) : null,
+        traceIdHigh ? span.traceId().substring(16) : span.traceId(),
+        span.parentId(),
+        span.id(),
+        span.kind() != null ? span.kind().name() : null,
+        span.name(),
+        span.timestamp(),
+        span.duration(),
+        span.localEndpoint() != null ? new EndpointUDT(span.localEndpoint()) : null,
+        span.remoteEndpoint() != null ? new EndpointUDT(span.remoteEndpoint()) : null,
+        annotations,
+        span.tags(),
+        annotation_query,
+        Boolean.TRUE.equals(span.debug()),
+        Boolean.TRUE.equals(span.shared())
+      );
```

Comment:
Retouching on the null binding resulting in tombstone issue.

Here there's 8 columns here that can be null. (maybe more?)

If there's consistently 8 tombstones (nulls) per row, then we'll only need 125 spans in a trace (rows in a partition) to trigger the `tombstone_warn_threshold warnings being logged in the C* nodes. And if we go to 12500 spans in a trace then that whole trace partition would become unreadable. Cassandra warns at a 1000 tombstones in any query, and fails on 100000 tombstones.

There's also a small question about disk usage efficiency. Each tombstone is a cell name and basically empty cell value entry stored on disk. Given that the cells are, apart from tags and annotations, generally very small then this could be proportionally an unnecessary waste of disk.

To avoid this relying upon a number of variant prepared statements for inserting a span is the normal practice.

Another popular practice is to insert those potentially null columns as separate statements (and optionally put them together into UNLOGGED batches). This works as multiple writes to the same partition has little overhead, and here we're not worried about lack of isolation between those writes, as the write is asynchronous anyway. An example of this approach is in the cassandra-reaper project here: https://github.com/thelastpickle/cassandra-reaper/blob/master/src/server/src/main/java/io/cassandrareaper/storage/CassandraStorage.java#L622-L642

Reply:
thanks captured